### ISSUE TITLE
Feature request: consider how to better support -split-sections

### ISSUE THREAD

**Issue by duog** [commercialhaskell/stack#3467]
-split-sections is a ghc flag new in 8.2.1:
https://downloads.haskell.org/~ghc/8.2.1/docs/html/users_guide/phases.html?highlight=split-sections#ghc-flag--split-sections

Unfortunately it is a bit of a pain to use this with stack because it requires one to compile all of one's dependencies with -split-sections.

I have created a stack.yaml and build script for stack itself here:
https://github.com/duog/stack/tree/split-sections

which builds stack from an empty stack root both with and without -split-sections. If one examines the build.out file in that repository one observes that there is very little difference in build time, but the resulting executable is about half the size with -split-sections. It's not demonstrated here, but the .a library files are much bigger.

As I'm sure you're aware, it is a bit difficult to ensure all the libraries in one's snapshot have certain ghc-options, is there anything that can be done to make this easier?

**Comment by mgsloan** [comment]
This sounds like the sort of thing where there isn't much downside.  May make sense to enable -split-sections by default.

**Comment by duog** [comment]
I've no doubt that enabling -split-sections by default is the right long-term solution, but it's a new feature and you'd need a way to reliably turn it off.

As devil's advocate, the downsides I'm aware of are:
* Not supported on windows in 8.2, and (currently) causes very slow linking in 8.4. See [trac:12913](https://ghc.haskell.org/trac/ghc/ticket/12913), [trac:13939](https://ghc.haskell.org/trac/ghc/ticket/13939)
* Seems to be the culprit in [trac:14291](https://ghc.haskell.org/trac/ghc/ticket/14291), I believe this only affects statically linked ghcs
* A bad default if you're producing dynamically linked executables, the .a archives are bloated for no benefit.

Though I've no doubt you've thought about it, it seems to me that the cabal-install new-* approach is the right one; a single package database with all the different flavours of packages. You could even have a per-snapshot database, though I don't see much upside to that. The current interface for specifying ghc-options in stack.yaml is great, it's just the interaction with already installed packages that causes me problems.

**Comment by mgsloan** [comment]
I haven't had much reason to use the cabal-install new-* stuff.  If I understand it correctly, one consequence is that you can't really load dependencies into ghci without fully specifying which package ids ghci should use.  With stack's approach, all of the packages that have been build for a particular snapshot / local db are available and consistent.

Once implemented, implicit snapshots will provide the benefits of both approaches.  It may have a little bit of overhead in creating new package DBs, but that seems to be really quick.  So, I think it is better to avoid the "everything in one DB" approach.

**Comment by duog** [comment]
Yes that's a good point re ghci. I think I was assuming stack would provide me a sub-database with only the packages from my stack.yaml.

I admit that I don't really follow exactly what an implicit snapshot would be, but I suppose it's something close to the sub-database I was thinking of.

The primary benefit of a single DB is in maximizing sharing between projects, hopefully implicit snapshots will provide this.

**Comment by mgsloan** [comment]
> The primary benefit of a single DB is in maximizing sharing between projects, hopefully implicit snapshots will provide this.

Yup, this can also be provided by multiple DBs.  Stack already has package sharing between snapshot DBs.  If some other snapshot DB already has the package (and it has the same dependencies), then it will just get registered in the other DB rather than rebuilt.

The thing that implicit snapshots adds to this is the possibility of also sharing extra-deps / possibly even git dependencies.

**Comment by domenkozar** [comment]
Seems to be default anyway since GHC 8.2.x on linux/darwin: https://ghc.haskell.org/trac/ghc/ticket/11445

**Comment by fosskers** [comment]
Possible relevant data. Binary sizes when compiling [Aura](https://github.com/aurapm/aura) with various resolvers "as-is":

-  `lts-11.16` (ghc 8.2.2) -> 28.6mb
- `nightly-2018-07-04` (ghc 8.4.3) -> 25.7mb

**Comment by Berengal** [comment]
I fell into this rabbit-hole of executable-size optimization today, and gathered some more data.

I tested three stack projects:
  - [hbfc](https://github.com/Berengal/hbfc/tree/c02efdd0b5eaaebeb0d44b8546a979dbf060a71d), my own project
  - [stack](https://github.com/commercialhaskell/stack/tree/da6c1a4616db074802fe7c14153b47c725fe6f87) itself
  - [aura](https://github.com/aurapm/aura/tree/831604b4daa5cd2699c4eb5ab95e21e02ad554d4), because it was tested previously.

On every project I ran `stack build`, checked the size of the executables in `.stack-work/dist/...`, ran `strip` on them and compared the size of the stripped executable to the one in `.stack-work/install/...` to make sure they matched (they did in all cases). I then added `"$everything": -split-sections` to the `ghc-options:` section in the `stack-yaml` file, adding the section if it wasn't there already, and repeated the same build and check.

The result:

  | executable | -split-sections | strip | size |
  |------------|:-----------------:|:-------:|------:|
  | hbfc       |       no        |  no   |  20M |
  | hbfc       |       no        |  yes  |  11M |
  | hbfc       |       yes       |  no   | 6,6M |
  | hbfc       |       yes       |  yes  | 3,6M |
  |            |                 |       |      |
  | hbfi       |       no        |  no   | 3,3M |
  | hbfi       |       no        |  yes  | 903K |
  | hbfi       |       yes       |  no   | 3,3M |
  | hbfi       |       yes       |  yes  | 903K |
  |            |                 |       |      |
  | stack      |       no        |  no   |  99M |
  | stack      |       no        |  yes  |  65M |
  | stack      |       yes       |  no   |  51M |
  | stack      |       yes       |  yes  |  33M |
  |            |                 |       |      |
  | aura       |       no        |  no   |  44M |
  | aura       |       no        |  yes  |  28M |
  | aura       |       yes       |  no   |  14M |
  | aura       |       yes       |  yes  | 8,4M |

Adding `-split-sections` to everything seems to make a big difference. The one exception is the `hbfi` executable in hbfc. I assume the reason for that it only uses `base` and the `hbfc` library, that `base` is already built with `-split-sections`, and that the library in the same package either builds with `-split-sections` automatically, or GHC optimizes away the unused parts anyway.

**Comment by fosskers** [comment]
Nice! See also: https://github.com/NixOS/nixpkgs/issues/43795

**Comment by sluukkonen** [comment]
As a data point, enabling split sections reduced the size of the binary from 50MB to 14MB in one of my personal projects.

**Comment by fosskers** [comment]
@sluukkonen did you do that with `stack`?

**Comment by fosskers** [comment]
@Berengal I'm going to test that myself with Aura right now.

**Comment by sluukkonen** [comment]
@fosskers Yes, I simply added

```
ghc-options:
  $everything: -split-sections
```

to my stack.yml.

**Comment by fosskers** [comment]
@sluukkonen Thanks, I'm going to try that myself.

**Comment by fosskers** [comment]
Yup, 8.8mb! I'm going to go ahead with this approach and make all my releases this way. Thank you!

**Comment by fosskers** [comment]
How about a `stack build --release` flag that forces `$everything: -split-sections`?

**Comment by sluukkonen** [comment]
I'm not sure if that would be ideal. 

Split sections requires all dependencies to built with it, so having a separate flag would mean that dependencies would be compiled twice.

Perhaps stack could switch them on by default at some point, but I'm not sure what the tradeoffs are. Compile times would be slower, at the very least.

**Comment by fosskers** [comment]
That would be the point of `--release`. Rust's `cargo` has such a feature, and it does indeed recompile everything.
